File: app/src/main/java/com/darkrockstudios/apps/adventurephotoframe/HealthMonitorService.java
package com.darkrockstudios.apps.adventurephotoframe;

import android.app.job.JobParameters;
import android.app.job.JobService;
import android.util.Log;

import retrofit2.Call;
import retrofit2.Response;

public class HealthMonitorService extends JobService
{
	private static final String TAG = HealthMonitorService.class.getSimpleName();

	public static final int  JOB_ID_PERIODIC = 1;
	public static final int  JOB_ID_INITIAL  = 2;
	public static final long JOB_PERIOD      = 24L * 60L * 60L * 1000L;

	@Override
	public boolean onStartJob( JobParameters params )
	{
		Log.i( TAG, "Checking in with Health Monitor..." );

		Call<Void> call = App.get().getNetworking().m_photoFrameService.healthCheckIn( App.get().getPhotoFrameId(),
		                                                                               BuildConfig.VERSION_CODE );
		call.enqueue( new Callback( params ) );

		return true;
	}

	@Override
	public boolean onStopJob( JobParameters params )
	{
		return true;
	}

	private class Callback implements retrofit2.Callback<Void>
	{
		private final JobParameters m_params;

		Callback( JobParameters params )
		{
			m_params = params;
		}

		@Override
		public void onResponse( Call<Void> call, Response<Void> response )
		{
			Log.i( TAG, "Health Check in complete" );
			jobFinished( m_params, false );
		}

		@Override
		public void onFailure( Call<Void> call, Throwable t )
		{
			Log.w( TAG, "Health Check in failed." );
			jobFinished( m_params, true );
		}
	}
}

File: app/src/main/java/com/darkrockstudios/iot/adventurephotoframe/healthmonitor/HealthMonitorService.kt
package com.darkrockstudios.iot.adventurephotoframe.healthmonitor

import android.app.job.JobParameters
import android.app.job.JobService
import android.util.Log
import com.darkrockstudios.iot.adventurephotoframe.BuildConfig
import com.darkrockstudios.iot.adventurephotoframe.application.App
import retrofit2.Call
import retrofit2.Response

class HealthMonitorService : JobService()
{
	override fun onStartJob(params: JobParameters): Boolean
	{
		if (!BuildConfig.DEBUG)
		{
			Log.i(TAG, "Checking in with Health Monitor...")

			val call = App.inst.networking.m_photoFrameService.healthCheckIn(App.inst.photoFrameId,
																			 BuildConfig.VERSION_CODE,
																			 "")
			call.enqueue(Callback(params))
		}

		return true
	}

	override fun onStopJob(params: JobParameters): Boolean
	{
		return true
	}

	private inner class Callback internal constructor(private val m_params: JobParameters) : retrofit2.Callback<Void>
	{

		override fun onResponse(call: Call<Void>, response: Response<Void>)
		{
			Log.i(TAG, "Health Check in complete")
			jobFinished(m_params, false)
		}

		override fun onFailure(call: Call<Void>, t: Throwable)
		{
			Log.w(TAG, "Health Check in failed.")
			jobFinished(m_params, true)
		}
	}

	companion object
	{
		private val TAG = HealthMonitorService::class.java.simpleName

		val JOB_ID_PERIODIC = 1
		val JOB_ID_INITIAL = 2
		val JOB_PERIOD = 24L * 60L * 60L * 1000L
	}
}
